src/evaluators/UnaryExpression.js

```diff
@@ -95,43 +94,26 @@ function evaluateDeleteOperation(expr: Value | Reference, realm: Realm) {
   return new BooleanValue(realm, bindings.DeleteBinding(referencedName));
 }
 
-function generateRuntimeCall(ref: Reference | Value, ast: BabelNodeUnaryExpression, strictCode: boolean, realm: Realm) {
-  invariant(ref instanceof Reference);
-  let baseValue = Environment.GetBase(realm, ref);
-  invariant(baseValue instanceof Value);
-  let propertyName = Environment.GetReferencedName(realm, ref);
-  invariant(typeof propertyName === "string");
-
-  if (!baseValue.isSimpleObject()) {
-    Leak.leakValue(realm, baseValue, ast.loc);
-  }
-  return AbstractValue.createTemporalFromBuildFunction(realm, Value, [baseValue], nodes => {
-    let arg;
-    if (typeof propertyName === "string") {
-      arg = t.isValidIdentifier(propertyName)
-        ? t.memberExpression(nodes[0], t.identifier(propertyName), false)
-        : t.memberExpression(nodes[0], t.stringLiteral(propertyName), true);
-    } else {
-      arg = t.memberExpression(nodes[0], nodes[1], true);
-    }
-    return t.unaryExpression(ast.operator, arg, ast.prefix);
-  });
+function generateRuntimeCall(expr: Value | Reference, ast: BabelNodeUnaryExpression, realm: Realm) {
+  let value = Environment.GetValue(realm, expr);
+  invariant(value instanceof AbstractValue);
+  return AbstractValue.createFromUnaryOp(realm, ast.operator, value);
 }
 
 function tryToEvaluateOperationOrLeaveAsAbstract(
   ast: BabelNodeUnaryExpression,
   expr: Value | Reference,
-  func: (expr: Value | Reference, realm: Realm) => Value,
+  func: (ast: BabelNodeUnaryExpression, expr: Value | Reference, strictCode: boolean, realm: Realm) => Value,
   strictCode: boolean,
   realm: Realm
 ) {
   let effects;
   try {
-    effects = realm.evaluateForEffects(() => func(expr, realm));
+    effects = realm.evaluateForEffects(() => func(ast, expr, strictCode, realm));
   } catch (error) {
     if (error instanceof FatalError) {
       return realm.evaluateWithPossibleThrowCompletion(
-        () => generateRuntimeCall(expr, ast, strictCode, realm),
+        () => generateRuntimeCall(expr, ast, realm),
```

Comment:
You seem to believe that func will throw a FatalError if and only if expr is either an abstract value or a reference to an abstract value. That is an intricate and subtle precondition to push to callers of tryToEvaluateOperationOrLeaveAsAbstract. Do you really need to do that?